src/domain/task-system/src/aggregates/task.rs

```diff
@@ -0,0 +1,276 @@
+    pub fn cancel(&mut self) -> Result<(), IllegalSequenceError<Self>> {
+        if self.cancellation_requested {
+            return Ok(());
+        }
+
+        self.apply(TaskCancelled {
+            event_time: Utc::now(),
+            task_id: self.task_id,
+        })
+    }
+
+    /// Transition task to a `Finished` state with the specified outcome
+    pub fn finish(&mut self, outcome: TaskOutcome) -> Result<(), IllegalSequenceError<Self>> {
+        self.apply(TaskFinished {
+            event_time: Utc::now(),
+            task_id: self.task_id,
+            outcome,
+        })
+    }
+
+    fn apply(&mut self, event: impl Into<TaskEvent>) -> Result<(), IllegalSequenceError<Self>> {
+        let event = event.into();
+        self.update_state(event.clone())?;
+        self.pending_events.push(event);
+        Ok(())
+    }
+
+    fn update_state(&mut self, event: TaskEvent) -> Result<(), IllegalSequenceError<Self>> {
+        assert_eq!(self.state.task_id, event.task_id());
+
+        // Check if state transition is legal
+        match (self.state.status, &event) {
+            (TaskStatus::Queued, TaskEvent::Running(_)) => {}
+            (TaskStatus::Queued | TaskStatus::Running, TaskEvent::Cancelled(_)) => {}
+            (TaskStatus::Queued | TaskStatus::Running, TaskEvent::Finished(_)) => {}
+            (_, _) => return Err(IllegalSequenceError::new(self.state.clone(), event).into()),
+        }
+
+        // Apply transition
+        match event {
+            TaskEvent::Created(_) => unreachable!(),
+            TaskEvent::Running(TaskRunning {
+                event_time,
+                task_id: _,
+            }) => {
+                self.state.status = TaskStatus::Running;
+                self.state.ran_at = Some(event_time);
+            }
+            TaskEvent::Cancelled(TaskCancelled {
+                event_time,
+                task_id: _,
+            }) => {
+                self.state.cancellation_requested = true;
+                self.state.cancellation_requested_at = Some(event_time);
+            }
+            TaskEvent::Finished(TaskFinished {
+                event_time,
+                task_id: _,
+                outcome,
+            }) => {
+                self.state.status = TaskStatus::Finished(outcome);
+                self.state.finished_at = Some(event_time);
+            }
+        }
+        Ok(())
+    }
+}
+
+/////////////////////////////////////////////////////////////////////////////////////////
+
+#[async_trait::async_trait]
+impl Aggregate for Task {
+    type Id = TaskID;
+    type Event = TaskEvent;
+    type State = TaskState;
+
+    fn id(&self) -> &TaskID {
+        &self.task_id
+    }
+
+    fn from_genesis_event(
+        event_id: EventID,
+        event: TaskEvent,
+    ) -> Result<Self, IllegalGenesisError<Self>> {
+        if !event.is_variant::<TaskCreated>() {
+            return Err(IllegalGenesisError { event });
+        }
```

Comment:
Why is this a dynamic check instead of better static type?

Reply:
I guess I could express this function as:
```rust
fn from_genesis_event<E>(
        event_id: EventID,
        event: E,
    ) -> Result<Self, IllegalGenesisError<Self>>
    where
        E: VariantOf<Self::Event>;
```

This would allow `Task::from_genesis_event` accept only `TaskCreated`.

BUT

This also limits aggregate (projection) to only initialize from a single type of event. I feel like in some cases it may be useful to allow multiple (or even all) event types to be able to define an initial state of projection.